### PR TITLE
[Serializer] add a context key to return always as collection for XmlEncoder

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #25227
| License       | MIT
| Doc PR        | 

This PR add a new `as_collection` context key in order to return always as a collection instead of returning a single elements when there are only one array.

there are only one PR for the CsvEncoder don't wanted to have only One PR containing the two changes. It feel better to have two PR that fix the behaviour on two different things. it's easy to review and to revert if it break something (which should not since we are testing the behaviour).

### CODE REVIEW

**Comment by Simperfit** [comment]
Travis failure is unrelated.

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 1]
```diff
@@ -37,17 +37,22 @@ class XmlEncoder implements EncoderInterface, DecoderInterface, NormalizationAwa
...
+    private $alwaysCollection;
 
     /**
      * Construct new XmlEncoder and allow to change the root node element name.
      *
      * @param string   $rootNodeName
      * @param int|null $loadOptions  A bit field of LIBXML_* constants
      */
-    public function __construct(string $rootNodeName = 'response', int $loadOptions = null)
+    public function __construct(string $rootNodeName = 'response', int $loadOptions = null/*, bool $alwaysCollection = false */)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: args can be appended to constructors, no need for the "BC layer"

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 2]
```diff
@@ -257,7 +262,6 @@ private function parseXml(\DOMNode $node, array $context = array())
 
         if (1 === count($value) && key($value)) {
             $data[key($value)] = current($value);
-
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: ?

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 3]
```diff
@@ -515,6 +515,39 @@ public function testDecodeIgnoreWhiteSpace()
         $this->assertEquals($expected, $this->encoder->decode($source, 'xml'));
     }
 
+    /**
+     * @group collection
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: why adding this?

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 4]
```diff
@@ -515,6 +515,39 @@ public function testDecodeIgnoreWhiteSpace()
...
+    public function testDecodeAlwaysAsCollection()
+    {
+        $this->encoder = new XmlEncoder('response', null, true);
+        $serializer = new Serializer(array(new CustomNormalizer()), array('xml' => new XmlEncoder()));
+        $this->encoder->setSerializer($serializer);
+
+        $source = <<<'XML'
+<?xml version="1.0"?>
+ <order_rows nodeType="order_row" virtualEntity="true">
+                <order_row>
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: indentation is strange, is it normal?

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 5]
```diff
@@ -515,6 +515,39 @@ public function testDecodeIgnoreWhiteSpace()
...
+            '@nodeType' => 'order_row',
+            '@virtualEntity' => 'true',
+            'order_row' => array(array(
+                'id' => 16,
+                'test' => 16
+            )),
+        );
+
+        $this->assertEquals($expected, $this->encoder->decode($source, 'xml'));
+    }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: missing blank line after

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]

**Comment by Simperfit** [comment]
AppVeyor failure is not related.

**Comment by Simperfit** [comment]
Status: Needs Review

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php (hunk 5 above)
Comment (Simperfit) [reply src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: added

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 6]
```diff
@@ -37,17 +37,18 @@ class XmlEncoder implements EncoderInterface, DecoderInterface, NormalizationAwa
...
 
     /**
      * Construct new XmlEncoder and allow to change the root node element name.
      *
-     * @param string   $rootNodeName
-     * @param int|null $loadOptions  A bit field of LIBXML_* constants
+     * @param int|null $loadOptions A bit field of LIBXML_* constants
      */
-    public function __construct(string $rootNodeName = 'response', int $loadOptions = null)
+    public function __construct(string $rootNodeName = 'response', int $loadOptions = null, bool $alwaysCollection = false)
```
Comment (fabpot) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: `$forceCollection`?

**Comment by dunglas** [comment]
@Simperfit can you rebase  to force the CI to run again?

**Comment by Simperfit** [comment]
done @dunglas

**Comment by Simperfit** [comment]
This is ready to be merged.

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 7]
```diff
@@ -335,7 +336,7 @@ private function parseXmlValue(\DOMNode $node, array $context = array())
         }
 
         foreach ($value as $key => $val) {
-            if (is_array($val) && 1 === count($val)) {
+            if (is_array($val) && 1 === count($val) && (!$this->forceCollection || !is_array($val[0]))) {
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: After a second thought, wouldn't it be better to allow to change this behavior through the context instead of forcing to register a new instance of the class?

Comment (Simperfit) [reply src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: I think it will be easier to use if we don't create a new instance. Ill do the modification on both PRs

**Comment by Simperfit** [comment]
@dunglas it now use a context key.

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 8]
```diff
@@ -333,9 +332,8 @@ private function parseXmlValue(\DOMNode $node, array $context = array())
                 $value[$subnode->nodeName][] = $val;
             }
         }
-
         foreach ($value as $key => $val) {
-            if (is_array($val) && 1 === count($val)) {
+            if (is_array($val) && 1 === count($val) && (!(isset($context['forceCollection']) && $context['forceCollection']) || !is_array($val[0]))) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: should use snake_case also I suppose

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 9]
```diff
@@ -333,9 +332,8 @@ private function parseXmlValue(\DOMNode $node, array $context = array())
                 $value[$subnode->nodeName][] = $val;
             }
         }
-
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: let's keep this line :)

**Comment by Simperfit** [comment]
done @nicolas-grekas

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 10]
```diff
@@ -41,8 +41,7 @@ class XmlEncoder implements EncoderInterface, DecoderInterface, NormalizationAwa
     /**
      * Construct new XmlEncoder and allow to change the root node element name.
      *
-     * @param string   $rootNodeName
-     * @param int|null $loadOptions  A bit field of LIBXML_* constants
+     * @param int|null $loadOptions A bit field of LIBXML_* constants
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: let's revert this: it's unrelated, and breaks merges

File: src/Symfony/Component/Serializer/Encoder/XmlEncoder.php [hunk 11]
```diff
@@ -335,7 +334,7 @@ private function parseXmlValue(\DOMNode $node, array $context = array())
         }
 
         foreach ($value as $key => $val) {
-            if (is_array($val) && 1 === count($val)) {
+            if (is_array($val) && 1 === count($val) && (!(isset($context['as_collection']) && $context['as_collection']) || !is_array($val[0]))) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: using `empty()` would make the code simpler I suppose
I don't understand the last condition `!is_array($val[0])` - it feels unrelated. Is it?

Comment (Simperfit) [reply src/Symfony/Component/Serializer/Encoder/XmlEncoder.php]: I'm looking for it, `!is_array($val[0])` makes it possible to return the value directly instead of having two arrays: 

```
$expected = array(
            '@nodeType' => 'order_row',
            '@virtualEntity' => 'true',
            'order_row' => array(array(
                'id' => 16,
                'test' => 16,
            )),
        );
```
vs
```
$expected = array(
            '@nodeType' => 'order_row',
            '@virtualEntity' => 'true',
            'order_row' => array(array(
                'id' => array(16),
                'test' => array(16),
            )),
        );
```

Do you mean replacing 
```
is_array($val) && 1 === count($val)
```

to 

```
!empty($val)
```

it's not possible

**Comment by Simperfit** [comment]
fabbot error will not be fixed in this PR since it's unrelated.

**Review by nicolas-grekas — APPROVED** [review]
Please update the main description so that people looking at this PR can understand what's going on here.
Please also check if a doc PR should be done.

**Comment by Simperfit** [comment]
@nicolas-grekas done.

**Comment by dunglas** [comment]
Thank you @Simperfit.
